ksqldb-execution/src/main/java/io/confluent/ksql/execution/codegen/CodeGenRunner.java

```diff
@@ -185,10 +188,29 @@ public Void visitFunctionCall(final FunctionCall node, final TypeContext context
       final List<SqlArgument> argumentTypes = new ArrayList<>();
       final FunctionName functionName = node.getName();
       for (final Expression argExpr : node.getArguments()) {
-        process(argExpr, null);
-        argumentTypes.add(SqlArgument.of(
-            expressionTypeManager.getExpressionSqlType(argExpr)
-        ));
+        process(argExpr, context);
+        final SqlType newSqlType = expressionTypeManager.getExpressionSqlType(argExpr, context);
+        // for lambdas - if we see this it's the  array/map being passed in we save the type
+        if (context.notAllInputsSeen()) {
+          if (newSqlType instanceof SqlArray) {
+            final SqlArray inputArray = (SqlArray) newSqlType;
+            context.addLambdaInputType(inputArray.getItemType());
+          } else if (newSqlType instanceof SqlMap) {
+            final SqlMap inputMap = (SqlMap) newSqlType;
+            context.addLambdaInputType(inputMap.getKeyType());
+            context.addLambdaInputType(inputMap.getValueType());
+          } else {
+            context.addLambdaInputType(newSqlType);
```

Comment:
I feel like we can put all of this into a new method in `TypeContext` called `context#visitType`:
```java
void visitType(SqlType type) {
  if (notAllInputsSeen()) {
      if (newSqlType instanceof SqlArray) {
          final SqlArray inputArray = (SqlArray) type;
          addLambdaInputType(inputArray.getItemType());
      } else if (newSqlType instanceof SqlMap) {
         final SqlMap inputMap = (SqlMap) type;
         addLambdaInputType(inputMap.getKeyType());
         addLambdaInputType(inputMap.getValueType())
      } else {
        addLambdaInputType(type);
    }
  }
}
```